sources/api/pluto/src/main.rs

```diff
@@ -144,42 +163,31 @@ async fn get_max_pods(client: &mut ImdsClient) -> Result<String> {
     error::NoInstanceTypeMaxPods { instance_type }.fail()
 }
 
-/// Returns the cluster's DNS IPV4 address. First it attempts to call EKS describe-cluster to find
-/// the `serviceIPv4CIDR`. If that works, it returns the expected cluster DNS IP address which is
-/// obtained by substituting `10` for the last octet. If the EKS call is not successful, it falls
-/// back to using IMDS MAC CIDR blocks to return one of two default addresses.
+/// Returns the cluster's DNS address.
+///
+/// For IPv4 clusters, first it attempts to call EKS describe-cluster to find the `serviceIpv4Cidr`.
+/// If that works, it returns the expected cluster DNS IP address which is obtained by substituting
+/// `10` for the last octet. If the EKS call is not successful, it falls back to using IMDS MAC CIDR
+/// blocks to return one of two default addresses.
+///
+/// For IPv6 clusters, EKS currently is choosing a random unique local address (ULA) for clusterIPs.
+/// Cluster DNS IP is `serviceIpv6Cidr` with the last hextet set to `a`
 async fn get_cluster_dns_ip(client: &mut ImdsClient) -> Result<String> {
-    // try calling eks describe-cluster to figure out the dns cluster ip
-    if let Some(dns_ip) = get_dns_from_eks().await {
-        // we were able to calculate the dns ip from the cidr range we received from eks
-        return Ok(dns_ip);
-    }
-
-    // we were unable to obtain or parse the cidr range from eks, fallback to one of two default
-    // values based on the cidr range of our primary network interface
-    get_cluster_dns_from_imds_mac(client).await
-}
-
-/// Gets the Service IPV4 CIDR setting from EKS and parses it to calculate the cluster DNS IP.
-/// Prints the error and returns `None` if anything goes wrong.
-async fn get_dns_from_eks() -> Option<String> {
-    let aws_k8s_info = match api::get_aws_k8s_info().await {
-        Ok(value) => value,
-        Err(e) => {
-            eprintln!(
-                "Unable to get region and cluster name from Bottlerocket API, using default DNS IP: {}",
-                e
-            );
-            return None;
+    // Retrieve the kubernetes network configuration for the EKS cluster
+    let aws_k8s_info = api::get_aws_k8s_info().await.context(error::AwsInfo)?;
+    let cluster_network_config =
+        eks::get_cluster_network_config(&aws_k8s_info.region, &aws_k8s_info.cluster_name)
+            .await
+            .context(error::EksError)?;
+    // Derive cluster-dns-ip from the service IPv4 CIDR
+    if let Some(ipv4_cidr) = cluster_network_config.service_ipv_4_cidr {
+        if let Ok(dns_ip) = get_dns_from_ipv4_cidr(&ipv4_cidr) {
+            return Ok(dns_ip);
         }
```

Comment:
The suggestion may not render well, but it looks like an error with EKS call would prevent fallback. I think you can use `if let Ok(cluster_network_config) {}` to ensure fallback.

(I could be wrong about this interpretation)

```suggestion
if let Ok(cluster_network_config) =
    eks::get_cluster_network_config(&aws_k8s_info.region, &aws_k8s_info.cluster_name)
        .await
        .context(error::EksError) {
    // Derive cluster-dns-ip from the service IPv4 CIDR
    if let Some(ipv4_cidr) = cluster_network_config.service_ipv_4_cidr {
        if let Ok(dns_ip) = get_dns_from_ipv4_cidr(&ipv4_cidr) {
            return Ok(dns_ip);
        }
    }
}
```

Reply:
Nice catch!